File: frontend/src/types/reports.ts
/**
 * Report types
 */
import { ApiResponse } from './common';

/**
 * Report Type
 */
export type ReportType =
  | 'performance'
  | 'risk'
  | 'optimization'
  | 'scenario'
  | 'historical'
  | 'comparison'
  | 'comprehensive'
  | 'custom';

/**
 * Report Format
 */
export type ReportFormat =
  | 'pdf'
  | 'html'
  | 'excel'
  | 'json'
  | 'csv';

/**
 * Report Frequency
 */
export type ReportFrequency =
  | 'daily'
  | 'weekly'
  | 'monthly'
  | 'quarterly'
  | 'annual'
  | 'once';

/**
 * Report Request
 */
export interface ReportRequest {
  portfolioId: string;
  reportType: ReportType;
  startDate?: string;
  endDate?: string;
  benchmark?: string;
  sections?: string[];
  format: ReportFormat;
  runInBackground?: boolean;
}

/**
 * Report Response
 */
export interface ReportResponse {
  status: 'completed' | 'processing' | 'failed';
  reportPath?: string;
  message?: string;
  portfolioId: string;
  reportType: ReportType;
  format: ReportFormat;
  generatedAt?: string;
  reportId?: string;
}

/**
 * Comparison Report Request
 */
export interface ComparisonReportRequest {
  portfolioIds: string[];
  reportType: ReportType;
  startDate?: string;
  endDate?: string;
  benchmark?: string;
  sections?: string[];
  format: ReportFormat;
  runInBackground?: boolean;
}

/**
 * Schedule Report Request
 */
export interface ScheduleReportRequest {
  portfolioId: string;
  reportType: ReportType;
  frequency: ReportFrequency;
  email?: string;
  reportParams: Record<string, any>;
}

/**
 * Schedule Report Response
 */
export interface ScheduleReportResponse {
  reportId: string;
  portfolioId: string;
  reportType: ReportType;
  frequency: ReportFrequency;
  status: 'active' | 'paused';
  nextRun?: string;
  email?: string;
}

/**
 * Report History Response
 */
export interface ReportHistoryResponse {
  reportId: string;
  portfolioId: string;
  reportType: ReportType;
  format: ReportFormat;
  generatedAt: string;
  status: 'completed' | 'failed';
  filePath?: string;
  fileSize?: number;
  durationMs?: number;
  parameters: Record<string, any>;
}

/**
 * Report Section
 */
export interface ReportSection {
  id: string;
  name: string;
  description?: string;
  required: boolean;
  default: boolean;
  order: number;
  dependencies?: string[];
}

/**
 * Report Template Request
 */
export interface ReportTemplateRequest {
  name: string;
  description?: string;
  sections: string[];
  format: ReportFormat;
  styleSettings?: Record<string, any>;
}

/**
 * Report Template Response
 */
export interface ReportTemplateResponse {
  templateId: string;
  name: string;
  description?: string;
  format: ReportFormat;
  sections?: ReportSection[];
  createdAt?: string;
  updatedAt?: string;
  isSystem: boolean;
}

/**
 * Report File Info Response
 */
export interface ReportFileInfoResponse {
  reportId: string;
  filePath: string;
  fileName: string;
  format: ReportFormat;
  mimeType: string;
  fileSize: number;
  createdAt: string;
  downloadUrl: string;
}

/**
 * Report Template
 */
export interface ReportTemplate {
  id: string;
  name: string;
  description: string;
  reportType: ReportType;
  sections: string[];
  format: ReportFormat;
  styleSettings: Record<string, any>;
  createdAt: string;
}

/**
 * Generated Report
 */
export interface Report {
  id: string;
  name: string;
  description: string;
  templateId?: string;
  portfolioId?: string;
  comparisonPortfolioIds?: string[];
  reportType: ReportType;
  format: ReportFormat;
  filePath?: string;
  startDate?: string;
  endDate?: string;
  benchmark?: string;
  sections?: string[];
  generationStatus: 'pending' | 'in_progress' | 'completed' | 'failed';
  metadata?: Record<string, any>;
  createdAt: string;
}

/**
 * Scheduled Report
 */
export interface ScheduledReport {
  id: string;
  name: string;
  description: string;
  templateId?: string;
  portfolioId?: string;
  comparisonPortfolioIds?: string[];
  reportType: ReportType;
  format: ReportFormat;
  frequency: ReportFrequency;
  nextGeneration?: string;
  lastGenerated?: string;
  emailRecipients?: string[];
  isActive: boolean;
  parameters?: Record<string, any>;
  createdAt: string;
}

export interface PortfolioCreateRequest {
  name: string;
  description?: string;
  type: string;
  benchmarkSymbol?: string;
  assets: Asset[];
  tags?: string[];
}

export enum PortfolioType {
  CONSERVATIVE = 'conservative',
  MODERATE = 'moderate',
  AGGRESSIVE = 'aggressive',
  CUSTOM = 'custom'
}

/**
 * API Report Response
 */
export type ApiReportResponse = ApiResponse<ReportResponse>;

/**
 * API Schedule Report Response
 */
export type ApiScheduleReportResponse = ApiResponse<ScheduleReportResponse>;

/**
 * API Report History Response
 */
export type ApiReportHistoryResponse = ApiResponse<ReportHistoryResponse[]>;

/**
 * API Report Templates Response
 */
export type ApiReportTemplatesResponse = ApiResponse<ReportTemplateResponse[]>;

/**
 * API Report Template Response
 */
export type ApiReportTemplateResponse = ApiResponse<ReportTemplateResponse>;

File: backend/app/api/endpoints/reports.py
from fastapi import APIRouter, Depends, HTTPException, Query, BackgroundTasks, Response
from typing import List, Optional, Dict, Any
from fastapi.responses import FileResponse
import os

from app.core.services.report_service import ReportService
from app.schemas.report import (
    ReportRequest,
    ReportResponse,
    ComparisonReportRequest,
    ScheduleReportRequest,
    ScheduleReportResponse,
    ReportHistoryResponse,
    ReportTemplateRequest,
    ReportTemplateResponse
)

router = APIRouter(prefix="/reports", tags=["reports"])


# Dependency to get the report service
def get_report_service():
    return ReportService()


@router.post("/generate", response_model=ReportResponse)
def generate_report(
        request: ReportRequest,
        background_tasks: BackgroundTasks,
        report_service: ReportService = Depends(get_report_service)
):
    """
    Generate a portfolio report.

    This endpoint creates a comprehensive report for a portfolio, including performance
    analysis, risk metrics, and other analytics. Reports can be generated in various formats.
    """
    try:
        # For very complex reports, we might want to run this in the background
        # and notify the client when it's done
        if request.run_in_background:
            # Add to background tasks
            background_tasks.add_task(
                report_service.generate_report,
                portfolio_id=request.portfolio_id,
                report_type=request.report_type,
                start_date=request.start_date,
                end_date=request.end_date,
                benchmark=request.benchmark,
                sections=request.sections,
                format=request.format
            )

            return {
                "status": "processing",
                "message": "Report generation started in background",
                "portfolio_id": request.portfolio_id,
                "report_type": request.report_type,
                "format": request.format
            }
        else:
            # Generate the report synchronously
            report_path = report_service.generate_report(
                portfolio_id=request.portfolio_id,
                report_type=request.report_type,
                start_date=request.start_date,
                end_date=request.end_date,
                benchmark=request.benchmark,
                sections=request.sections,
                format=request.format
            )

            return {
                "status": "completed",
                "report_path": report_path,
                "portfolio_id": request.portfolio_id,
                "report_type": request.report_type,
                "format": request.format
            }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Report generation failed: {str(e)}")


@router.post("/compare", response_model=ReportResponse)
def generate_comparison_report(
        request: ComparisonReportRequest,
        background_tasks: BackgroundTasks,
        report_service: ReportService = Depends(get_report_service)
):
    """
    Generate a portfolio comparison report.

    This endpoint creates a report comparing multiple portfolios, highlighting differences
    in composition, performance, risk characteristics, and other metrics.
    """
    try:
        if request.run_in_background:
            # Add to background tasks
            background_tasks.add_task(
                report_service.generate_comparison_report,
                portfolio_ids=request.portfolio_ids,
                report_type=request.report_type,
                start_date=request.start_date,
                end_date=request.end_date,
                benchmark=request.benchmark,
                sections=request.sections,
                format=request.format
            )

            return {
                "status": "processing",
                "message": "Comparison report generation started in background",
                "portfolio_ids": request.portfolio_ids,
                "report_type": request.report_type,
                "format": request.format
            }
        else:
            # Generate the report synchronously
            report_path = report_service.generate_comparison_report(
                portfolio_ids=request.portfolio_ids,
                report_type=request.report_type,
                start_date=request.start_date,
                end_date=request.end_date,
                benchmark=request.benchmark,
                sections=request.sections,
                format=request.format
            )

            return {
                "status": "completed",
                "report_path": report_path,
                "portfolio_ids": request.portfolio_ids,
                "report_type": request.report_type,
                "format": request.format
            }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Comparison report generation failed: {str(e)}")


@router.post("/schedule", response_model=ScheduleReportResponse)
def schedule_report(
        request: ScheduleReportRequest,
        report_service: ReportService = Depends(get_report_service)
):
    """
    Schedule a periodic report generation.

    This endpoint sets up automated generation of reports on a regular schedule,
    with options for delivery via email or other notification methods.
    """
    try:
        report_id = report_service.schedule_report(
            portfolio_id=request.portfolio_id,
            report_type=request.report_type,
            frequency=request.frequency,
            email=request.email,
            **request.report_params
        )

        return {
            "report_id": report_id,
            "portfolio_id": request.portfolio_id,
            "report_type": request.report_type,
            "frequency": request.frequency,
            "status": "scheduled"
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to schedule report: {str(e)}")


@router.get("/scheduled", response_model=List[ScheduleReportResponse])
def list_scheduled_reports(
        report_service: ReportService = Depends(get_report_service)
):
    """
    List all scheduled reports.

    This endpoint returns a list of all reports that are scheduled for automatic generation.
    """
    try:
        scheduled_reports = report_service.list_scheduled_reports()
        return scheduled_reports
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to list scheduled reports: {str(e)}")


@router.delete("/scheduled/{report_id}", status_code=204)
def cancel_scheduled_report(
        report_id: str,
        report_service: ReportService = Depends(get_report_service)
):
    """
    Cancel a scheduled report.

    This endpoint stops the automatic generation of a previously scheduled report.
    """
    try:
        success = report_service.cancel_scheduled_report(report_id)
        if not success:
            raise HTTPException(status_code=404, detail=f"Scheduled report '{report_id}' not found")
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to cancel scheduled report: {str(e)}")


@router.get("/history", response_model=List[ReportHistoryResponse])
def list_report_history(
        portfolio_id: Optional[str] = None,
        report_type: Optional[str] = None,
        start_date: Optional[str] = None,
        end_date: Optional[str] = None,
        report_service: ReportService = Depends(get_report_service)
):
    """
    List report generation history.

    This endpoint returns a history of generated reports, with optional filtering
    by portfolio, report type, and date range.
    """
    try:
        history = report_service.list_report_history(
            portfolio_id=portfolio_id,
            report_type=report_type,
            start_date=start_date,
            end_date=end_date
        )
        return history
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to list report history: {str(e)}")


@router.get("/templates", response_model=List[ReportTemplateResponse])
def get_report_templates(
        report_service: ReportService = Depends(get_report_service)
):
    """
    Get available report templates.

    This endpoint lists all available templates that can be used for report generation.
    """
    try:
        templates = report_service.get_report_templates()
        return templates
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get report templates: {str(e)}")


@router.post("/templates", response_model=ReportTemplateResponse)
def create_report_template(
        request: ReportTemplateRequest,
        report_service: ReportService = Depends(get_report_service)
):
    """
    Create a custom report template.

    This endpoint allows users to define and save custom templates for future report generation.
    """
    try:
        template_id = report_service.create_report_template(
            name=request.name,
            description=request.description,
            sections=request.sections,
            format=request.format,
            style_settings=request.style_settings
        )

        return {
            "template_id": template_id,
            "name": request.name,
            "description": request.description,
            "format": request.format
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to create report template: {str(e)}")


@router.get("/download/{report_id}")
def download_report(
        report_id: str,
        report_service: ReportService = Depends(get_report_service)
):
    try:

        reports = report_service.list_report_history(report_id=report_id)
        if not reports:
            raise HTTPException(status_code=404, detail=f"Report '{report_id}' not found")

        report_info = reports[0]
        if 'file_path' not in report_info:
            raise HTTPException(status_code=404, detail=f"Report file path not found")

        file_path = report_info['file_path']

        if not os.path.exists(file_path):
            raise HTTPException(status_code=404, detail=f"Report file not found")

        file_name = os.path.basename(file_path)

        return FileResponse(
            path=file_path,
            filename=file_name,
            media_type=report_service.get_content_type(file_path)
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to download report: {str(e)}")